examples/keras_io/nlp/neural_machine_translation_with_keras_nlp.py

```diff
@@ -0,0 +1,515 @@
+    )
+    return vocab
+
+
+"""
+Every vocabulary has a few special, reserved tokens. We have four such tokens:
+
+- `"[PAD]"` - Padding token. Padding tokens are appended to the input sequence
+length when the input sequence length is shorter than the maximum sequence length.
+- `"[UNK]"` - Unknown token.
+- `"[START]"` - Token that marks the start of the input sequence.
+- `"[END]"` - Token that marks the end of the input sequence.
+"""
+
+reserved_tokens = ["[PAD]", "[UNK]", "[START]", "[END]"]
+
+eng_samples = [text_pair[0] for text_pair in train_pairs]
+eng_vocab = train_word_piece(eng_samples, ENG_VOCAB_SIZE, reserved_tokens)
+
+spa_samples = [text_pair[1] for text_pair in train_pairs]
+spa_vocab = train_word_piece(spa_samples, SPA_VOCAB_SIZE, reserved_tokens)
+
+"""
+Let's see some tokens!
+"""
+
+print("English Tokens: ", eng_vocab[100:110])
+print("Spanish Tokens: ", spa_vocab[100:110])
+
+"""
+Now, let's define the tokenizers. We will configure the tokenizers with the
+the vocabularies trained above.
+"""
+
+eng_tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=eng_vocab, lowercase=False
+)
+spa_tokenizer = keras_nlp.tokenizers.WordPieceTokenizer(
+    vocabulary=spa_vocab, lowercase=False
+)
+
+"""
+Let's try and tokenize a sample from our dataset! To verify whether the text has
+been tokenized correctly, we can also detokenize the list of tokens back to the
+original text.
+"""
+
+eng_input_ex = text_pairs[0][0]
+eng_tokens_ex = eng_tokenizer.tokenize(eng_input_ex)
+print("English sentence: ", eng_input_ex)
+print("Tokens: ", eng_tokens_ex)
+print(
+    "Recovered text after detokenizing: ",
+    eng_tokenizer.detokenize(eng_tokens_ex),
+)
+
+print()
+
+spa_input_ex = text_pairs[0][1]
+spa_tokens_ex = spa_tokenizer.tokenize(spa_input_ex)
+print("Spanish sentence: ", spa_input_ex)
+print("Tokens: ", spa_tokens_ex)
+print(
+    "Recovered text after detokenizing: ",
+    spa_tokenizer.detokenize(spa_tokens_ex),
+)
+
+"""
+## Format datasets
+
+Next, we'll format our datasets.
+
+At each training step, the model will seek to predict target words N+1 (and beyond)
+using the source sentence and the target words 0 to N.
+
+As such, the training dataset will yield a tuple `(inputs, targets)`, where:
+
+- `inputs` is a dictionary with the keys `encoder_inputs` and `decoder_inputs`.
+`encoder_inputs` is the tokenized source sentence and `decoder_inputs` is the target
+sentence "so far",
+that is to say, the words 0 to N used to predict word N+1 (and beyond) in the target
+sentence.
+- `target` is the target sentence offset by one step:
+it provides the next words in the target sentence -- what the model will try to predict.
+
+We will add special tokens, `"[START]"` and `"[END]"`, to the input Spanish
+sentence after tokenizing the text. We will also pad the input to a fixed length.
+This can be easily done using `keras_nlp.layers.StartEndPacker`.
+"""
+
+
+def preprocess_batch(eng, spa):
+    batch_size = spa.shape[0]
```

Comment:
I'm surprised this works -- I'd assume this needs to be `ops.shape(spa)[0]`

Reply:
I can change it to `ops.shape(spa)[0]`. This works because spa is a tf tensor (to process data from tf.data).

Reply:
Let's use `ops.shape(spa)[0]` for better generality.

Reply:
done